### PR TITLE
Refs [TASK][26257] Create DAO

### PR DESCRIPTION
**Pull request by quockhanhng**
## Related Tickets
- [#26257](https://edu-redmine.sun-asterisk.vn/issues/26257)

## WHAT
- Create DAO

## HOW
- I create TaskDAO and ListDAO interface and then create classes implement them

## WHY
- To communicate with local database

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Reviewer1 (KhanhNQ) | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAO.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.sunasterisk.getitdone.data.source.local.dao
+
+import com.sunasterisk.getitdone.data.model.TaskList
+
+interface ListDAO {
```
Comment (tuannm-1276) [inline app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAO.kt]: list về gì em? đặt tên lại interface và các phương thức con. tương tự ở nơi khác.
Nên đặt tên dễ hiểu

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAOImpl.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import android.database.Cursor
+import com.sunasterisk.getitdone.data.model.TaskList
+import com.sunasterisk.getitdone.data.source.local.database.AppDatabase
+
+class ListDAOImpl private constructor(appDatabase: AppDatabase) : ListDAO {
+
+    private val database = appDatabase.writableDatabase
+
+    override fun getAllLists(): List<TaskList> {
+        val cursor = database.query("LIST", null, null, null, null, null, null)
```
Comment (tuannm-1276) [inline app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAOImpl.kt]: LIST, TITLE, ... hardcode

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAO.kt (hunk 1 above)
Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAO.kt]:
```suggestion
interface TasksDAO {
```

Comment (quockhanhng) [reply app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/ListDAO.kt]: vậy em sửa thành TaskList cho giống tên model ạ

File: app/src/main/java/com/sunasterisk/getitdone/data/model/Task.kt [hunk 3]
```diff
@@ -17,6 +20,19 @@ data class Task(
...
+        cursor.getInt(cursor.getColumnIndex(Database.TABLE_TASK_COL_ID)),
+        cursor.getInt(cursor.getColumnIndex(Database.TABLE_TASK_COL_LIST_ID)),
+        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_TITLE)),
+        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_DESCRIPTION)),
+        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_TIME_REMINDER)),
+        cursor.getInt(cursor.getColumnIndex(Database.TABLE_TASK_COL_IS_IMPORTANT)) == TRUE,
+        cursor.getInt(cursor.getColumnIndex(Database.TABLE_TASK_COL_IS_IN_MY_DAY)) == TRUE,
+        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_COLOR)),
+        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_STATUS)),
+        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_TIME_CREATED))
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/getitdone/data/model/Task.kt]: Các colum này để luông trong class này luôn Khánh 
```suggestion
        cursor.getInt(cursor.getColumnIndex(ID)),
        cursor.getInt(cursor.getColumnIndex(LIST_ID)),
        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_TITLE)),
        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_DESCRIPTION)),
        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_TIME_REMINDER)),
        cursor.getInt(cursor.getColumnIndex(Database.TABLE_TASK_COL_IS_IMPORTANT)) == TRUE,
        cursor.getInt(cursor.getColumnIndex(Database.TABLE_TASK_COL_IS_IN_MY_DAY)) == TRUE,
        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_COLOR)),
        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_STATUS)),
        cursor.getString(cursor.getColumnIndex(Database.TABLE_TASK_COL_TIME_CREATED))
```

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    override fun getAllTasksByListId(listId: Int): List<Task> {
+        return getTaskInSpecificCondition(Database.TABLE_TASK_COL_LIST_ID, listId.toString())
+    }
+
+    override fun getTaskById(id: Int): Task {
+        val cursor =
+            database.query(
+                Database.TABLE_TASK_NAME,
+                null,
+                Database.TABLE_TASK_COL_ID,
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt]:
```suggestion
                Task.TABLE_NAME,
                null,
                Task.ID,
```

File: app/src/main/java/com/sunasterisk/getitdone/data/model/Task.kt (hunk 3 above)
Comment (quockhanhng) [reply app/src/main/java/com/sunasterisk/getitdone/data/model/Task.kt]: vâng ạ

File: app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            put("LIST_ID", task.listId)
+            put("TITLE", task.title)
+            put("DESCRIPTION", task.description)
+            put("TIME_REMINDER", task.timeReminder)
+            put("IS_IMPORTANT", task.isImportant)
+            put("IS_IN_MY_DAY", task.isInMyDay)
+            put("COLOR", task.color)
+            put("STATUS", task.status)
+            put("TIME_CREATED", task.timeCreated)
+        }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/getitdone/data/source/local/dao/TaskDAOImpl.kt]: Chuyển vào trong model luôn 
```suggestion
        val values = task.getContentValues()
```

**Pull request merged**